sdk/trace/sampling.go

```diff
@@ -64,56 +68,258 @@ type SamplingResult struct {
+	var low int
+	if has := strings.HasPrefix(otts, "rv:"); has {
+		low = 3
+	} else if pos := strings.Index(otts, ";rv:"); pos > 0 {
+		low = pos + 4
+	} else {
+		return 0, false
+	}
+	if len(otts) < low+14 {
+		otel.Handle(fmt.Errorf("could not parse tracestate randomness: %q: %w", otts, strconv.ErrSyntax))
+	} else if len(otts) > low+14 && otts[low+14] != ';' {
+		otel.Handle(fmt.Errorf("could not parse tracestate randomness: %q: %w", otts, strconv.ErrSyntax))
+	} else {
+		randomIn := otts[low : low+14]
+		if rv, err := strconv.ParseUint(randomIn, 16, 64); err == nil {
+			randomness = rv
+			hasRandom = true
+		} else {
+			otel.Handle(fmt.Errorf("could not parse tracestate randomness: %q: %w", randomIn, err))
+		}
+	}
+	return
 }
 
 func (ts traceIDRatioSampler) ShouldSample(p SamplingParameters) SamplingResult {
 	psc := trace.SpanContextFromContext(p.ParentContext)
-	x := binary.BigEndian.Uint64(p.TraceID[8:16]) >> 1
-	if x < ts.traceIDUpperBound {
+	state := psc.TraceState()
+
+	existOtts := state.Get("ot")
+
+	var randomness uint64
+	var hasRandom bool
+	if existOtts != "" {
+		// When the OTel trace state field exists, we will
+		// inspect for a "rv", otherwise assume that the
+		// TraceID is random.
+		randomness, hasRandom = tracestateHasRandomness(existOtts)
+	}
+	if !hasRandom {
+		// Interpret the least-significant 8-bytes as an
+		// unsigned number, then zero the top 8 bits using
+		// randomnessMask, yielding the least-significant 56
+		// bits of randomness, as specified in W3C Trace
+		// Context Level 2.
+		randomness = binary.BigEndian.Uint64(p.TraceID[8:16]) & randomnessMask
+	}
+	if ts.threshold > randomness {
 		return SamplingResult{
-			Decision:   RecordAndSample,
-			Tracestate: psc.TraceState(),
+			Decision:   Drop,
+			Tracestate: state,
 		}
 	}
+
+	if mod, err := state.Insert("ot", combineTracestate(existOtts, ts.otts)); err == nil {
+		state = mod
+	} else {
+		otel.Handle(fmt.Errorf("could not update tracestate: %q", err))
+	}
 	return SamplingResult{
-		Decision:   Drop,
-		Tracestate: psc.TraceState(),
+		Decision:   RecordAndSample,
+		Tracestate: state,
 	}
 }
 
+// combineTracestate combines an existing OTel tracestate fragment,
+// which is the value of a top-level "ot" tracestate vendor tag.
+func combineTracestate(incoming, updated string) string {
+	// `incoming` is formatted according to the OTel tracestate
+	// spec, with colon separating two-byte key and value, with
+	// semi-colon separating key-value pairs.
+	//
+	// `updated` should be a single two-byte key:value to modify
+	// or insert therefore colonOffset is 2 bytes, valueOffset is
+	// 3 bytes into `incoming`.
+	const colonOffset = 2
+	const valueOffset = colonOffset + 1
+
+	if incoming == "" {
+		return updated
+	}
+	var out strings.Builder
+
+	// The update is expected to be a single key-value of the form
+	// `XX:value` for with two-character key.
+	upkey := updated[:colonOffset]
```

Comment:
I see that this is guaranteed to have a `th:` prefix in the current code, but do you think it would be worth having a sanity check to future-proof it?